AA-Contract/account-abstraction
# Socialrecovery

### Implementation
1. Register guardian
   - Only owner can register guardian
2. Confirm recovery
   - Only guardian registered can vote to recovery wallet
5. Recovery wallet
   - When recovery tokens above threshold are accumulated in the wallet, the owner is changed
4. Delete guardian
   - Only owner can delete guardian
---

### For Security
1. Only the Guardian, which has passed a certain period(at least a day) of time since the guardian registration, can vote to recover wallet
3. After deleting the Guardian, the voting will not be effective until a certain period(at least a day) of time has passed and owner can cancel the deletion
6. Can not register guardian above max supply
7. Can not transfer the recovery token to anyone except for a registered recovery wallet
---
### For Utility
Record the recovery history for each guardian. We can know who is trustworthy as a guardian

--- 
### Test Case

1. Register guardian (Mint Token)
    1. Change of totalSupply 
    8. Revert when exceed maxSupply 
    9. Revert when guardian duplicated
2. Delete guardian (Burn token)
    1. Change of totalSupply
    2. Revert before time delay
    6. Delete from Guardian List after time deplay
10. Recovery wallet
    1. Balance change of recovery wallet
    2. Change of owner when the confirmation is more than a threshold
    3. Update nonce after recovery wallet
    11. Revert about duplicate confirm (vote)
    12. Revert when caller is not a guardian
    13. Revert before time delay
7. Deploy Social Recovery wallet through userOp
